feat(whale-api): add GET api for on-chain governance

kyleleow opened the pull request:
<!--  Thanks for sending a pull request! -->

#### What this PR does / why we need it:
- Add `/proposals`, `/proposals/:id` and `/proposals/:id/votes` endpoints on whale
- Add `.listGovProposals()`, `.getGovProposal()`, and `listGovProposalVotes()` to `WhaleApiClient`
- To allow straightforward paginated listing implementation for downstream, e.g. scan and wallet

#### Which issue(s) does this PR fixes?:
<!--
(Optional) Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
Fixes #

#### Additional comments?:
- Intermittent codecov issue when running CI, seems to be related to https://github.com/codecov/codecov-action/issues/598

kyleleow committed:
feat: add get api to whale

kyleleow committed:
feat: add e2e for get OCG api

Fix default masternode for listGovProposalVotes

Add ProposalVoteResultType enum

Linting

kyleleow committed:
feat: update e2e and enum for proposal vote

To align with ain 3.2.0 fixes

kyleleow committed:
Merge branch 'main' into kengye/governance-whale-api

kyleleow committed:
Merge branch 'main' into kengye/governance-whale-api

kyleleow committed:
fix: update to match latest listgovproposals rpc

kyleleow committed:
Merge remote-tracking branch 'upstream/main' into kengye/governance-whale-api

kyleleow committed:
chore: typo on function helper

kyleleow committed:
fix: update controller to match latest rpc

Add more e2e to cover all param

kyleleow committed:
chore: just to trigger CI

kyleleow committed:
Revert "chore: just to trigger CI"

This reverts commit b7fdfc70be708c99fcbc0edf0f4a0125c46eaf07.

kyleleow committed:
chore: to rerun CI

kyleleow committed:
Revert "chore: to rerun CI"

Skip test due to blockchain issue

kyleleow committed:
chore: add tests to return all records

kyleleow committed:
Merge remote-tracking branch 'upstream/main' into kengye/governance-whale-api

kyleleow committed:
chore: remove .only

kyleleow committed:
test: fix broken tests

By moving listProposalVotes into a separate test suite

kyleleow committed:
chore: test if push reflect on GH PR

kyleleow committed:
Revert "chore: test if push reflect on GH PR"

This reverts commit 8d8c66eb8f6463fd1806d233e2745f514f8106e0.

kyleleow committed:
Merge branch 'main' into kengye/ocg-whale-api

eli-lim commented on apps/whale-api/src/module.api/governance.controller.e2e.ts:
```diff
@@ -0,0 +1,328 @@
+      masternodeId = id
+    }
+  }
+  return masternodeId
```
Throw an error if this is still `''`

kyleleow replied:
updated

eli-lim commented on apps/whale-api/src/module.api/governance.controller.e2e.ts:
```diff
@@ -0,0 +1,328 @@
+  // Listing related tests
+  it('should listProposals', async () => {
+    const result = await controller.listProposals()
+    expect(result.data.length).toStrictEqual(2)
```
should have at least one test that asserts the proposal shape? e.g. `title`, `context`, etc.
instead of just testing `result.data.length`

kyleleow replied:
updated, also added some manipulation because rpc itself doesnt support sorting, data returned is of random order

eli-lim commented on apps/whale-api/src/module.api/governance.controller.e2e.ts:
```diff
@@ -0,0 +1,328 @@
+
+  it('should listProposalVotes', async () => {
+    const result = await controller.listProposalVotes(cfpProposalId)
+    expect(result.data.length).toStrictEqual(3)
```
Same here - test the shape on top of the length

kyleleow replied:
updated

eli-lim commented on apps/whale-api/src/module.api/governance.service.ts:
```diff
@@ -0,0 +1,160 @@
+        err instanceof RpcApiError &&
+        (err?.payload?.message === `Proposal <${id}> not found` ||
+          err?.payload?.message ===
+            "proposalId must be of length 64 (not 3, for '999')")
```
Hmm? Where is this `(not 3, for '999')` from? `ain`?

kyleleow replied:
Ops removed, that part should be dynamic, updated to use `.includes()`

eli-lim commented on apps/whale-api/src/module.api/governance.service.ts:
```diff
@@ -0,0 +1,160 @@
+  ): Promise<ApiPagedResponse<ProposalVotesResult>> {
+    try {
+      const next = query.next !== undefined ? Number(query.next) : undefined
+      const size = query.size > 30 ? 30 : query.size
```
```suggestion
      const size = Math.min(query.size, 30)
```

`nit` :3

kyleleow replied:
updated

eli-lim commented on packages/jellyfish-api-core/src/category/governance.ts:
```diff
@@ -31,6 +31,13 @@ export enum VoteDecision {
+  YES = 'YES',
+  NO = 'NO',
+  NEUTRAL = 'NEUTRAL',
+  Unknown = 'Unknown'
```
Do we want to standardise with the other enums? Either `UPPERCASE`, `PascalCase` or `lowercase` ?

kyleleow replied:
Updated to use `UPPERCASE` for both key and values

eli-lim replied:
How about `Unknown`?

kyleleow replied:
on `jellyfish-api-core`, i am keeping `Unknown` to match what is returned from the `ain` side ([ref](https://github.com/DeFiCh/ain/blob/master/src/masternodes/proposals.cpp#L39-L48))
on `whale-api` i am mapping to return the uppercase `UNKNOWN`

eli-lim commented on apps/whale-api/src/module.api/governance.service.ts:
```diff
@@ -0,0 +1,160 @@
+    })
+  }
+
+  async get (id: string): Promise<GovernanceProposal> {
```
```suggestion
  async getGovernanceProposal (id: string): Promise<GovernanceProposal> {
```

`nit`

eli-lim replied:
https://github.com/JellyfishSDK/jellyfish/pull/1970#discussion_r1070856219

kyleleow replied:
updated

eli-lim commented on apps/whale-api/src/module.api/governance.service.ts:
```diff
@@ -0,0 +1,160 @@
+export class GovernanceService {
+  constructor (private readonly client: JsonRpcClient) {}
+
+  async list (
```
```suggestion
  async listGovernanceProposal (
```

eli-lim replied:
Unless we call this service `GovernanceProposalService`

kyleleow replied:
updated the method name, not renaming the file for now, will KIV if more methods are added in the future

eli-lim reviewed (commented):
Generally LGTM! Although I don't have full context

kyleleow committed:
Merge branch 'main' into kengye/ocg-whale-api

kyleleow committed:
chore: assert proposal shape

Fix function naming, enum casing, helper typo

Throw error when votable masternode not found during controller e2e

kyleleow committed:
Merge branch 'main' into kengye/ocg-whale-api

eli-lim commented on apps/whale-api/src/module.api/governance.controller.ts:
```diff
@@ -0,0 +1,64 @@
+  @Get('/proposals')
+  async listProposals (
+    @Query() status: ListProposalsStatus = ListProposalsStatus.ALL,
+    @Query() type: ListProposalsType = ListProposalsType.ALL,
```
Not sure if we want to implement any enum-based validation here

Since it's an enum in jellyfish, it's likely that we're expected to manually keep this in sync with `ain`?
Or is this some Whale-specific enum?

kyleleow replied:
added into `governance.test.ts` using nestjs built-in pipe at the query string level

eli-lim reviewed (commented):
Final comment

kyleleow commented:
Changing to draft while tests are being added into `packages/whale-api-client/__tests__`

eli-lim commented:
> Changing to draft while tests are being added into packages/whale-api-client/__tests__

@kyleleow Alternatively if we don't want to _manually keep it in sync with ain_ and there are no foreseeable problems passing any arbitrary string to defid, then we can skip the validation

But generally I'd be more conservative and add it 😛

kyleleow committed:
feat: add governance to whale api

todo: tests

kyleleow committed:
feat: add tests to OCG in WhaleApiClient

Use nestjs built-in pipe for enum validation and default value for param

Use `all` boolean flag to return all record, as opposed to size=0 design from ain

Add optional query param to requestList

kyleleow committed:
Merge remote-tracking branch 'upstream/main' into kengye/ocg-whale-api

kyleleow committed:
Merge branch 'main' into kengye/ocg-whale-api

kyleleow committed:
refactor: use object for OCG api endpoint for Whale

Use interface for Whale request param from jellyfish-api-core

eli-lim commented on packages/whale-api-client/__tests__/api/governance.test.ts:
```diff
@@ -0,0 +1,375 @@
+        message: 'Unable to find proposal',
+        url: '/v0.0/regtest/governance/proposals/123/votes?size=30&masternode=mine&cycle=0&all=false'
+      })
+    }
```
This will pass even if the listGovProposalVotes promise resolves

Should do something like this instead:

```typescript
await expect(
  client.governance.listGovProposalVotes({ id: '123' })
).rejects.toThrowError(
  new WhaleApiException(...)
)
```

eli-lim commented on packages/whale-api-client/__tests__/api/governance.test.ts:
```diff
@@ -0,0 +1,375 @@
+    expect(resultPage2.length).toStrictEqual(0)
+  })
+
+  it('should listProposals with error when using invalid status', async () => {
```
```suggestion
  it('should throw error when listProposals called with invalid status', async () => {
```

eli-lim commented on packages/whale-api-client/__tests__/api/governance.test.ts:
```diff
@@ -0,0 +1,375 @@
+        at: expect.any(Number),
+        message: 'Invalid query parameter value for status. See the acceptable values: voting, rejected, completed, all',
+        url: '/v0.0/regtest/governance/proposals?size=30&status=123&type=all&cycle=0&all=false'
+      })
```
same issue - https://github.com/JellyfishSDK/jellyfish/pull/1970#discussion_r1080850212

eli-lim commented on packages/whale-api-client/__tests__/api/governance.test.ts:
```diff
@@ -0,0 +1,375 @@
+    }
+  })
+
+  it('should listProposals with error when using invalid type', async () => {
```
```suggestion
  it('should throw error when listProposals called with invalid type', async () => {
```

eli-lim commented on packages/whale-api-client/__tests__/api/governance.test.ts:
```diff
@@ -0,0 +1,375 @@
+      // To skip typescript validation in order to assert invalid query parameter
+      // @ts-expect-error
+      await client.governance.listGovProposals({ type: '123' })
+    } catch (err) {
```
same issue - https://github.com/JellyfishSDK/jellyfish/pull/1970#discussion_r1080850212

fullstackninja864 committed:
updated testcase for listGovProposalVotes

fullstackninja864 committed:
Merge branch 'main' into kengye/ocg-whale-api

fullstackninja864 committed:
Merge branch 'main' into kengye/ocg-whale-api

eli-lim reviewed (approved):
I see the tests are duplicated between the controller.e2e and the client tests
Personally I'd prefer just having the client drive the tests, but I guess that would be part of some larger refactoring work

imkven reviewed (approved):
LGTM.
